README.md
<p align="center">
  <a href="https://d3portillo.github.io/ira/">
    <img width="420" src="./assets/ira.svg"/>
  </a>
</p>

---

  <strong>
  <a href="https://codepen.io/D3Portillo/pen/poyjxwM">📝 Codepen</a> ・
  <a href="https://observablehq.com/@d3portillo/ira-fetch-wrapper">🔬 Observable</a> ・
  <a href="https://www.npmjs.com/package/irajs">📦 Npm</a>
  </strong>

---

**Ira Fetch:** Vanilla JS Fetch API wrapper with goodies 🍒

Irajs is a **small ~ 3kb** function which enhances current Fetch API with more, more goodies. Ira code isn't chopped, replaced with random chars or similar on .min.js version, it's just minified.

Ira goodies include debug logs, persistent settings and custom currying to request functions with a set of options. The little wrapper tries to function using current JS Engine features, no babel or typescript used just plain vanilla Javascript.

![Version](https://img.shields.io/npm/v/irajs/latest)
![Size](https://img.shields.io/bundlephobia/minzip/irajs)
![Downloads](https://img.shields.io/npm/dw/irajs)

## Npm Install

```
npm install irajs
```

## Yarn Install

```
yarn add irajs
```

## CDN Load

```html
<script src="https://d3portillo.github.io/ira/src/index.min.js"></script>
```

## Usage

```js
import ira from "irajs"
// Requires
const ira = require("irajs")
// Start playing around
ira.get("/stuff")
```

[👉 The complete API reference](#Ira-API-Reference)

## Some examples

### Getting data

```js
ira.get(`https://postman-echo.com/get?foo1=bar1&foo2=bar2`).then(({ data }) => {
  console.log(data.json, data.text, data.blob) // * Automatic response parsing
})
```

### Adding headers

```js
ira.config({
  headers: {
    "Content-type": "application/json",
  },
})
```

### Parsing blob to base64 string

```js
const blob = new Blob(["something"])
ira.blobToBase64(blob).then((base64) => console.log(base64))
```

### Including a base URL

```js
const request = ira.extend({
  baseURL: "https://yourendpoint.com/dev/branch",
})
// Now you can do
request.get("/binary") //https://yourendpoint.com/dev/branch/binary
```

### Extending a fork

A custom settings fork of main Ira function that's gapped to provided - config

```js
const request = ira.extend({
  headers: {
    "x-api-key": "somsaltedencryptedawesomekey",
  },
  debug: true /* Shows Ira stuff on console */,
  parseBlob: false /* Do not include .blob on data */,
})

// Now you can make requests containing those settings
request
  .get("https://something")
  .then(({ data: { blob } }) => console.info(null == blob))
// The blob response inside data obj is null
```

> This method extends from Ira Object

## The Ira Instance

```js
RESPONSE = {
  data: { json: Object, text: String, blob: ?Blob }
  ok: Boolean,
  status: Number,
  statusText: String,
  statusCode: status<Number>,
  error: ?Error
}
ON_REQUEST_PROPS = {
  headers: {},
  body: ?String,
  debug: ?Boolean,
  parseBlob: ?Boolean,
  params: {},
  ...`https://developer.mozilla.org/en-US/docs/Web/API/Request`
}
IRA_SETTINGS = {
  headers: {},
  debug: Boolean,
  parseBlob: Boolean,
  baseURL: ?String,
}
HTTP_METHODS = {
  get: Function,
  put: Function,
  post: Function,
  head: Function,
  delete: Function,
  connect: Function,
  options: Function,
  trace: Function,
}

// Exported object { Main }
ira = {
  ...HTTP_METHODS,
  default(): HTTP_METHODS.get,
  _settings: Object,
  config: Function,
  extend: Function() => ira /* Fork with provided config */,
  blobToBase64: Function
}
```

Ira will return a void response if an error ocurred and status of 500.

## Ira API Reference

### Table of contents

- [ira()](#default)
- [ira.get()](#get)
- [ira.post()](#post)
- [ira.put()](#post)
- [ira.delete()](#delete)
- [ira.connect()](#connect)
- [ira.options()](#options)
- [ira.trace()](#trace)
- [ira.head()](#head)
- [ira.blobToBase64()](#blobtobase64)
- [ira.on()](#onevent)
- [ira.extend()](#extend)
- [ira.config()](#config)
- [ira.\_config](#getconfig)
- [{ baseURL }](#baseurl)
- [{ params }](#params)
- [{ debug }](#debug)
- [{ parseBlob }](#parseblob)

### HTTP Methods

This interface rules among many props and methods, **`ON_REQUEST_PROPS:` [ira#the-ira-instance](https://github.com/D3Portillo/ira#the-ira-instance)**

---

<a href="#default" id="default"># </a><b>ira</b>([<i>URL, CONFIG</i>]) [<>](https://github.com/D3Portillo/ira/blob/master/src/index.js#L41 "Source")

```js
URL = "" // Your endpoint URL
CONFIG = ON_REQUEST_PROPS
```

The GET method, `ira("/something")` is the same as `fetch("/something")`. The difference is ira.get returns a `Promise` which resolves to an Object including .json, .text and .blob methods.

`ira().then(({data}) => { data.json | data.text | data.blob })`

<a href="#get" id="get"># </a>ira<b>.get</b>([<i>URL, CONFIG</i>]) [<>](https://github.com/D3Portillo/ira/blob/master/src/index.js#L169 "Source")

```js
URL = "" // That stuff URL
CONFIG = ON_REQUEST_PROPS
```

Same as `ira()` method.

<a href="#post" id="post"># </a>ira<b>.post</b>([<i>URL, CONFIG</i>]) [<>](https://github.com/D3Portillo/ira/blob/master/src/index.js#L178 "Source")

```js
URL = "" // An endpoint
CONFIG = ON_REQUEST_PROPS
```

The POST method, `ira.post("/something")` is the same as `fetch("/something", { method: "POST" })`.

You can include a body doing:

```js
ira.post("/something", {
  body: "some-body",
})
```

<a href="#put" id="put"># </a>ira<b>.put</b>([<i>URL, CONFIG</i>]) [<>](https://github.com/D3Portillo/ira/blob/master/src/index.js#L179 "Source")

```js
URL = "" // https://something
CONFIG = ON_REQUEST_PROPS
```

The HTTP PUT method, `ira.put("/api")` is the same as `fetch("/api", { method: "PUT" })`.

You can include a body doing:

```js
ira.put("/something", {
  body: "some-body",
})
```

You also can show some debug messages on console by adding `debug: true`.

```js
ira.put("/something", {
  body: "some-body",
  debug: true,
})

// This will log on request start messages and
// When promise is resolved with your data
```

<a href="#delete" id="delete"># </a>ira<b>.delete</b>([<i>URL, CONFIG</i>]) [<>](https://github.com/D3Portillo/ira/blob/master/src/index.js#L175 "Source")

```js
URL = "" // Place an URL here
CONFIG = ON_REQUEST_PROPS
```

That DELETE Http Method, `ira.delete("/api")` is the same as `fetch("/api", { method: "DELETE" })`.

<a href="#connect" id="connect"># </a>ira<b>.connect</b>([<i>URL, CONFIG</i>]) [<>](https://github.com/D3Portillo/ira/blob/master/src/index.js#L180 "Source")

```js
URL = "" // The place you want data from
CONFIG = ON_REQUEST_PROPS
```

Doin a CONNECT method `ira.connect("/api")` is same as `fetch("/api", { method: "CONNECT" })`.

You can include this config on your request:

```js
{
  headers: {}, // Your request headers
  body: ?String, // Your request body
  debug: ?Boolean, // if true shows some stuff on console
  parseBlob: ?Boolean, // if false .blob event wont execute
  params: {} // Your URL params ?reqid=3&something=data
}
```

<a href="#options" id="options"># </a>ira<b>.options</b>([<i>URL, CONFIG</i>]) [<>](https://github.com/D3Portillo/ira/blob/master/src/index.js#L181 "Source")

```js
URL = "" // That URL you want data from
CONFIG = ON_REQUEST_PROPS
```

When doing the OPTIONS Http method `ira.options("/api")` results to be same as doing `fetch("/api", { method: "OPTIONS" })`.

<a href="#trace" id="trace"># </a>ira<b>.trace</b>([<i>URL, CONFIG</i>]) [<>](https://github.com/D3Portillo/ira/blob/master/src/index.js#L41 "Source")

```js
URL = "" // Production or dev endpoint
CONFIG = ON_REQUEST_PROPS
```

TRACE method, `ira.trace("/api")` is the same as `fetch("/api", { method: "TRACE" })`.

<a href="#head" id="head"># </a>ira<b>.head</b>([<i>URL, CONFIG</i>]) [<>](https://github.com/D3Portillo/ira/blob/master/src/index.js#L182 "Source")

```js
URL = "" // Some resource URL
CONFIG = ON_REQUEST_PROPS
```

The HEAD method, `ira.head("/api")` is the same as `fetch("/api", { method: "HEAD" })`.

### Yes, more cool methods

<a href="#blobtobase64" id="blobtobase64"># </a>ira<b>.blobToBase64</b>([<i>Blob</i>]) [<>](https://github.com/D3Portillo/ira/blob/master/src/index.js#L207 "Source")

```js
Blob = new Blob() // A JS Binary long object
```

You can parse any `Blob` into a base64 string by doing `ira.blobToBase64`. This returns a Promise which resolves into a String. This method will always resolve if there's an error check out you console. If Promise fails will resolve to `""`

<a href="#onevent" id="onevent"># </a>ira<b>.on</b>([<i>Event</i>]) [<>](https://github.com/D3Portillo/ira/blob/master/src/index.js#L233 "Source")

```js
Event = "request" | "response"
```

You can add a custom listener when ira or it's forks perform a request or when a response is succeded. You can set this triggers like this `ira.on("response", callback)`.

Example:

```js
ira.on("request", (request) => {
  const { url, statusCode, method, headers, config } = request
  if (statusCode == 200) {
    console.log("This will always succeed")
    /*
     This callback is made as soon as request is made
     so statusCode is 200, you can log config, check path and include some magic
    */
  }
})

ira.on("response", (response) => {
  const { url, statusCode, method, headers, config } = request
  // Lets say you want to kick user when it's forbidden
  if (statusCode == 403) killSession()
})
```

<a href="#extend" id="extend"># </a>ira<b>.extend</b>([<i>CONFIG</i>]) [<>](https://github.com/D3Portillo/ira/blob/master/src/index.js#L244 "Source")

```js
CONFIG = {
  headers: {},
  debug: Boolean,
  parseBlob: Boolean,
  baseURL: ?String,
} // @see https://github.com/D3Portillo/ira#the-ira-instance
```

This method returns a new Ira instance, you can replace default Settings with your custom ones. This can be helpfull if making request to API's where headers are somewhat "persistent", for example x-api-key's or that.

Example:

```js
const request = ira.extend({
  headers: {
    "x-api-key": "somethingawesome",
    "Content-type": "application/json",
  },
})

// Then you can avoid rewriting those headers again

request.get("/endpoint", {
  body: {
    user: "D3Portillo",
    base: "Somebass",
  },
})
// This will include those headers added on .extend call
```

<a href="#config" id="config"># </a>ira<b>.config</b>([<i>CONFIG</i>]) [<>](https://github.com/D3Portillo/ira/blob/master/src/index.js#L194 "Source")

```js
CONFIG = {
  headers: {},
  debug: Boolean,
  parseBlob: Boolean,
  baseURL: ?String,
} // @see https://github.com/D3Portillo/ira#the-ira-instance
```

This method is used to replace current ira or fork settings. This replaces .extend method stuff with those new ones you provide.

```js
const req = ira.extend({
  baseURL: "https://google.com",
})

// Now, let's pretend you want to change that baseURL
req.settings({
  baseURL: "https://duckduckgo.com",
})
// This will replace request's baseURL google.com with duckduck.go
```

### Acces current config

<a href="#getconfig" id="getconfig"># </a>ira<b>.\_config</b><i>\<CONFIG></i> [<>](https://github.com/D3Portillo/ira/blob/master/src/index.js#L188 "Source")

```js
CONFIG = {
  headers: {},
  debug: Boolean,
  parseBlob: Boolean,
  baseURL: ?String,
} // @see https://github.com/D3Portillo/ira#the-ira-instance
```

If you want to check current ira config do `ira._config`. This is supposed to be changed with `ira.config()`, still you can set `ira.\_config.headers = {}`

### Config[`._config`] props

<a href="#baseurl" id="baseurl"># </a>ira<b>.\_config.baseURL</b><i>\<Boolean></i> [<>](https://github.com/D3Portillo/ira/blob/master/src/index.js#L74 "Source")

You can add a baseURL to make your requests.

```js
ira.settings({
  baseURL: "https://localhost:5000",
})
ira.get("/anurl", {
  params: {
    somelikeparam: "somevalue",
  },
}) // Fetches https://localhost:5000/anurl
```

<a href="#params" id="params"># </a>ira<b>.\_config.params</b><i>\<Object></i> [<>](https://github.com/D3Portillo/ira/blob/master/src/index.js#L76 "Source")

You can add params on your request like this:

```js
ira.get("/anurl", {
  params: {
    somelikeparam: "somevalue",
  },
}) // Fetches to /anurl?somelikeparam=somevalue
```

<a href="#debug" id="debug"># </a>ira<b>.\_config.debug</b><i>\<Boolean></i> [<>](https://github.com/D3Portillo/ira/blob/master/src/index.js#L140 "Source")

If `true` will log stuff on console when a request is made and when a response is obtained.

<a href="#parseblob" id="parseblob"># </a>ira<b>.\_config.parseBlob</b><i>\<Boolean></i> [<>](https://github.com/D3Portillo/ira/blob/master/src/index.js#L105 "Source")

If `false` any request you make wont perform a **response.blob** and your data will resolve with this as `null`

### Some resources

- Source: [/src/index.js](./src/index.js)
- Changelog: [/CHANGELOG.md](./CHANGELOG.md)
- License: [/LICENSE](./LICENSE)

---

**Ira** stands for: `Go-to` in spanish `Ir-a`. Can also mean rage or anger, That's all the feelings you have while handling HTTP stuff : )


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [0.0.6] - 2020-08-30

### Added

- `on()` event listeners, currently functioning with on "request" and "response" events occur.

### Fixed

- `.text` method will return `""` if text is from a binary.

## [0.0.5] - 2020-08-16

### Changed

- Forks now return another Ira instance

### Fixed

- All `.blob, .text, .json` responses are provided now

## [0.0.4-beta0.2] - 2020-08-16

### Added

- Minified version - really added, comments kept : )
- **Now you can .extend.extend.extend.... Ira's:** Keep in mind that from a `.extend` function prop `._config` will return config provided on function call.

### Changed

- Custom function export has been replaced

### Fixed

- Removed string version of Blob if it's text type or type's undefined

## [0.0.4-beta0.1] - 2020-08-15

### Added

- You can now parse blob to base64 - `ira.blobToBase64`

### Fixed

- Now user wont get utf-8 version of binary inside data.blob

### Removed

- `.reset() method` , you can use `ira.config({})`

## [0.0.4] - 2020-08-13

### Added

- Params can be passed from request: `ira.get("/", { params: { } })`
- This log was created

### Changed

- Root const's were changed to var


test.js
console.info("GET METHOD")
ira.get(`https://postman-echo.com/get?foo1=bar1&foo2=bar2`).then(({ data }) => {
  console.log(data.json, data.text, data.blob)
  // Automatic response parsing
})

console.info("POST METHOD")
ira
  .post(`https://postman-echo.com/post`, {
    body: "This body will be returned",
    params: {
      token: 2,
    },
  })
  .then(({ data }) => {
    console.log(data.json, data.text, data.blob)
  })

console.info("GET METHOD With custom headers")
const request = ira.extend({
  headers: {
    "x-api-key": "BOg81b54cfD99ufEmd21sxgp696bm7XcT2F2jVOzdw21",
    session: "LsZoeYLx2cAwMsauPsP56nWgMpu3T89Jy",
  },
  debug: true,
  label: "Request with custom headers",
  baseURL: "https://documenter.getpostman.com/view/3632562377",
})

request.get("/something")
ira.get("/", { params: { foo1: "bar1", foo2: "bar2" } , debug: true})


tests/index.js
// * DO NOT INCLUDE require("ira") here
window.onload = () => {
  const init = () => {
    document.body.style = `font-family: Roboto, Open sans, Ubuntu, Tahoma; padding: 1rem`
    document.body.innerHTML = `
    <b style="font-size: 2rem">
      Content log...
    </b>
    <img src="http://localhost:5000/blob" alt="This will fail if Blob error response"/>
    <br/><br/>
    <i>Open da ~ Dev console</i>
    <hr style="height: 1px; border: none; background: rgba(0,0,0,0.35); margin: 1rem 0"/>
    <pre><code style="white-space: normal" id="logTag"></code></pre>
    `
    /**
     * Log to DOM Function
     * @param {?string} text
     * @param {?Boolean} isError
     */
    return (text = "") => {
      const log = window.logTag
      const t = JSON.stringify(text)
      const content = `<span>${t}</span>`
      log.insertAdjacentHTML("beforeend", `${content}<br/><br/>`)
      console.log(text)
    }
  }
  const log = init()

  //* ///////////////////////////////////////////////////////////////
  /**
   ** Test Playground for Ira - Fetch Wrapper
   */

  // * Response {data,text,blob} test

  ira.get("/text").then(({ data }) => {
    console.log("text", data)
    log(`TEXT_TEST: ${data.text != ""}`)
  })
  ira.get("/json").then(({ data }) => {
    console.log("json", data)
    log(`JSON_TEST: ${typeof data.json == "object"}`)
  })
  ira.get("/blob").then(({ data }) => {
    console.log("blob", data)
    log(`BLOB_TEST: ${data.blob != null}`)
  })

  // Extra stuff
  log("Get image")
  ira(`https://animeonegai.com/assets/img/spring4.png`).then(({ data }) => {
    ira.blobToBase64(data.blob).then((image) => {
      document.body.insertAdjacentHTML(
        "beforeend",
        `<h2>Base 64 Image</h2><img src="${image}"/>`
      )
    })
  })

  ira.config({
    headers: {
      "first-config": true,
    },
    debug: true,
  })
  log(ira._config)

  const second = ira.extend({
    headers: {
      second: true,
    },
  })
  second.on("request", (onCB) => console.error({ onCB }))
  second.get("/")
  second("/second")
  log(second._config)
  second.config({
    headers: {},
    baseURL: "/",
    debug: true,
  })
  log(second._config)
  const third = second.extend({
    headers: {
      one: 2,
    },
    baseURL: "/2",
  })
  third()
  log(third._config)
}


tests/panda.js
module.exports = `<svg width="90" height="109" viewBox="0 0 90 109" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M44.8653 68.7945C33.2779 68.7946 31.5251 78.2201 34.1061 91.1411C36.1007 95.9838 36.425 98.0239 35.7613 100.659C34.1315 106.17 32.0222 107.893 25.4157 107.694C20.4324 105.477 17.7777 103.274 13.4148 97.3484C4.2721 84.3157 1.47617 76.8913 1 63.4147C1.65894 47.6949 4.72912 39.7937 15.4839 27.4119C17.9991 26.1288 18.938 26.5377 20.036 28.6534C20.4496 39.1308 23.0359 43.8675 32.0369 50.1724C42.1832 55.461 48.0645 55.3991 58.9356 50.1724C65.8825 45.4417 68.36 40.7262 69.695 28.6534C71.3246 25.5677 72.5872 25.7637 75.0747 27.4119C88.2295 44.903 89.7859 53.2395 88.731 67.1392C84.9392 88.5014 80.586 94.0022 72.5918 103.142C67.607 107.959 64.9832 108.835 60.5908 107.694C55.0181 102.924 53.5134 99.5551 56.4526 91.1411C59.4463 76.9304 56.4526 68.7945 44.8653 68.7945Z" fill="black" stroke="black"/><path d="M62.2466 27.8256C62.2466 31.0253 59.2822 33.6192 55.6254 33.6192C51.9686 33.6192 49.0042 31.0253 49.0042 27.8256C48.9118 27.5947 48.9174 27.3689 49.0042 27.1489L61.2413 24.0115C61.9293 24.851 62.2466 26.208 62.2466 27.8256Z" fill="black"/><path d="M27.4848 27.6303C27.4848 30.9379 30.449 33.6193 34.1055 33.6193C37.7621 33.6193 40.7262 30.9379 40.7262 27.6303C40.8186 27.3916 40.8131 27.1582 40.7262 26.9307L28.4901 23.6875C27.8021 24.5553 27.4848 25.9581 27.4848 27.6303Z" fill="black"/><path d="M77.9715 10.8588C77.9715 16.5726 73.5248 21.2045 68.0396 21.2045C65.3355 15.1008 63.1345 12.4916 58.2344 9.20352C58.995 4.27663 63.0953 0.513184 68.0396 0.513184C73.5248 0.513184 77.9715 5.14508 77.9715 10.8588Z" fill="black"/><path d="M12.5872 10.3829C12.5872 15.8338 17.0344 20.2527 22.5202 20.2527C25.2247 14.4298 27.4259 11.9406 32.3267 8.80376C31.566 4.10352 27.4651 0.513184 22.5202 0.513184C17.0344 0.513184 12.5872 4.93201 12.5872 10.3829Z" fill="black"/><ellipse cx="44.8653" cy="40.6543" rx="6.62121" ry="4.55208" fill="black"/></svg>`


src/index.js
/**
 * Ira Fetch - Vanilla JS Fetch API wrapper with goodies 🍒
 * @author Denny Portillo<hello@d3portillo.me>
 * @license MIT
 * @see https://github.com/d3portillo/ira
 * @version 0.0.6
 */

function f(forkConfig = {}) {
  const IRA_CONFIG = {
    headers: {},
    debug: false,
    parseBlob: true,
    baseURL: undefined,
  }
  const IRA_METHODS = {
    get: "GET",
    put: "PUT",
    post: "POST",
    head: "HEAD",
    delete: "DELETE",
    connect: "CONNECT",
    options: "OPTIONS",
    trace: "TRACE",
  }
  const IRA_LOG = "IraFetch >>>"
  const deepify = (obj = {}) => {
    const json = {}
    Object.keys(obj).forEach((prop) => {
      const V = obj[prop]
      json[prop] = typeof V != "string" ? JSON.stringify(V) : V
    })
    return json
  }

  /**
   * Ira.get Method
   * @param { String } url
   * @param { IRA_CONFIG } extra
   */
  function ira(url = "/", extra) {
    return makeIraFetch(IRA_METHODS.get, {
      acceptsBody: false,
    })(url, extra)
  }
  function makeIraFetch(method = "POST", opt = { acceptsBody: true }) {
    /**
     * Ira Response Object
     * @typedef { Object } IraResponse
     * @property {{ json: Object, text: String, blob: ?Blob }} data - Posible parsed response body
     * @property { Boolean } ok - Response status <= 300
     * @property { Number } status
     * @property { String } statusText
     * @property { Number } statusCode
     * @property { ?Error } error - Null if nothing wrong
     */
    /**
     * @param { String } url - URL To fetch from
     * @param {{ headers: {}, body: ?String, params: {}, debug: Boolean, parseBlob: Boolean }} extra - Your normal fetch opts
     * @return { Promise<IraResponse> }
     */
    const fetchPromise = (url = "/", extra = {}) => {
      const config = {
        ...IRA_CONFIG,
        ...extra,
        ...forkConfig,
        ...ira._config,
      }
      let { headers = {}, body = "", params = {} } = extra
      headers = deepify({ ...config.headers, ...headers })
      const cType = headers["Content-Type"] || headers["Content-type"] || ""
      if (cType.includes("json")) body = deepify(body)
      try {
        let { baseURL, parseBlob } = config
        url = baseURL ? `${baseURL}${url}` : url
        const paramsChain = new URLSearchParams(params).toString()
        if (paramsChain.length) {
          const lastChar = url.charAt(url.length - 1)
          if (lastChar != "/") url = `${url}/`
          url = `${url}?${paramsChain}`
        }
        const { fetch } = window
        if (!fetch) throw new Error("Not inside a browser")
        if (!url) throw new Error("URL not provided")
        return new Promise((send) => {
          ira.__exec_on.request({
            url,
            statusCode: 200,
            method,
            headers: new Headers(headers),
            config,
          })
          if (config.debug) console.info(`${IRA_LOG} URL='${url}' >>> SENT ⚡`)
          fetch(url, {
            ...extra,
            method,
            headers,
            ...(opt.acceptsBody ? { body } : {}),
          })
            .then((response) => {
              const { ok, status, statusText } = response
              const clone = response.clone()
              Promise.all([
                response.text(),
                parseBlob ? clone.blob() : null,
              ]).then(([t, b]) => {
                if (typeof b == "string") {
                  ;[t, b] = [b, t]
                }
                let json = {}
                try {
                  json = JSON.parse(t)
                  if (typeof json == "string") json = {}
                } catch (_) {}
                if (parseBlob) {
                  const BIN_TYPE = /video|image|audio|ogg|pdf|rar|zip|font|7z|flash|x-/g
                  // Posible blob types to binaries - this will be too ugly parsed
                  // to utf-8 string from resposne.text
                  const isBinary = BIN_TYPE.test(b.type)
                  const isUTF8 = /utf8|utf-8/gi.test(b.type)
                  if (isBinary && !isUTF8) t = ""
                }
                const data = { json, text: t, blob: b }
                const statusCode = status
                send({
                  data,
                  ok,
                  status,
                  statusText,
                  statusCode,
                  error: null,
                })
                ira.__exec_on.response({
                  url,
                  statusCode,
                  method,
                  headers: response.headers,
                  config,
                })
                if (config.debug) {
                  console.info(`🍃 ${IRA_LOG} URL='${url}' >>> RESPONSE: `, {
                    config,
                    responseData: { data },
                  })
                }
              })
            })
            .catch((error) => {
              console.error(
                `⛔ ${IRA_LOG} - Got error on request, URL='${url}' >>> ${error}`,
                { config }
              )
              send({
                data: { json: {}, text: "", blob: null },
                ok: false,
                status: 500,
                statusText: error,
                statusCode: 500,
                error,
              })
            })
        })
      } catch (e) {
        console.error(`${IRA_LOG} >>> ${e}`)
      }
    }
    return fetchPromise
  }
  ira.get = makeIraFetch(IRA_METHODS.get, {
    acceptsBody: false,
  })
  ira.head = makeIraFetch(IRA_METHODS.head, {
    acceptsBody: false,
  })
  ira.delete = makeIraFetch(IRA_METHODS.delete, {
    acceptsBody: false,
  })
  ira.post = makeIraFetch()
  ira.put = makeIraFetch(IRA_METHODS.put)
  ira.connect = makeIraFetch(IRA_METHODS.connect)
  ira.options = makeIraFetch(IRA_METHODS.options)
  ira.trace = makeIraFetch(IRA_METHODS.trace)

  /**
   * Acces your current Ira config
   * @type { IRA_CONFIG }
   */
  ira._config = {}

  /**
   * Sets persisten config headers or body for future requests
   * @param { IRA_CONFIG } config
   */
  ira.config = (config = {}) => {
    if (!Object.entries(config).length) {
      // We reset config if *empty { } , is provided
      config = { ...IRA_CONFIG }
    }
    ira._config = { ...config }
  }

  /**
   * Returns a base64 String from a blob
   * @param { Blob } blob
   * @returns { Promise<String> }
   */
  ira.blobToBase64 = (blob) => {
    return new Promise((cb) => {
      try {
        const rdr = new FileReader()
        rdr.onload = () => cb(rdr.result)
        rdr.readAsDataURL(blob)
      } catch (e) {
        console.error(`${IRA_LOG} - Got error on Blob parse >>> ${e}`)
        cb("")
      }
    })
  }

  ira.__exec_on = {
    request: () => null,
    response: () => null,
  }
  /**
   * @callback cbFunction
   * @param {{ url: String, statusCode: ?Number, method: String, headers: {}, config: {} }} callbackObject
   */
  /**
   * Add your custom callbacks on response and request events
   * @param {("request" | "response")} event
   * @param { cbFunction } callback - Request or Response Objects Callback
   */
  ira.on = (event, callback) => {
    if (typeof callback == "function" && ira.__exec_on[event]) {
      ira.__exec_on[event] = callback
    }
  }

  /**
   * Sets config and returns a custom Ira Fork
   * @param { IRA_CONFIG } config
   * @returns { ira }
   */
  ira.extend = (conf = {}) => {
    conf.headers = conf.headers ? conf.headers : {}
    const iH = ira._config.headers
    conf.headers = iH ? { ...iH, ...conf.headers } : conf.headers
    const __ira = f(conf)
    __ira.config({ ...ira._config, ...conf })
    return __ira
  }
  return ira
}

"object" == typeof exports && "object" == typeof module
  ? (module.exports = f())
  : "object" == typeof exports
  ? (exports.ira = f())
  : (this.ira = f())
